server/src/services/organizations.service.ts

```diff
@@ -539,4 +551,101 @@ export class OrganizationsService {
+    return groups
+      .split('|')
+      .map((group: string) =>
+        group === 'All Users' || group === 'Admin' ? decamelize(group.replace(' ', '')) : group
+      );
+  }
+
+  async bulkUploadUsers(currentUser: User, fileStream, res: Response) {
+    const users = [];
+    const existingUsers = [];
+    const invalidRows = [];
+    const invalidGroups = [];
+    const emailPattern = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i;
+
+    csv
+      .parseString(fileStream.toString(), {
+        headers: ['first_name', 'last_name', 'email', 'groups'],
+        renameHeaders: true,
+        ignoreEmpty: true,
+      })
+      .transform((row: UserCsvRow, next) => {
+        return next(null, {
+          ...row,
+          groups: this.decamelizeDefaultGroupNames(row?.groups),
+        });
+      })
+      .validate(async (data: UserCsvRow, next) => {
+        await dbTransactionWrap(async (manager: EntityManager) => {
+          //Check for existing users
+          const user = await this.usersService.findByEmail(data?.email);
+          if (user?.organizationUsers?.some((ou) => ou.organizationId === currentUser.organizationId)) {
+            existingUsers.push(data?.email);
+          } else {
+            users.push(data);
+          }
+
+          //Check for invalid groups
+          const receivedGroups = data?.groups;
+          for (const group of receivedGroups) {
+            const orgGroupPermission = await manager.findOne(GroupPermission, {
+              where: {
+                organizationId: currentUser.organizationId,
+                group: group,
+              },
+            });
+
+            if (!orgGroupPermission) {
+              invalidGroups.push(group);
+            }
+          }
+          return next(null, data.first_name !== '' && data.last_name !== '' && emailPattern.test(data.email));
+        });
+      })
+      .on('data', function () {})
+      .on('data-invalid', (row, rowNumber) => {
+        invalidRows.push(rowNumber);
+      })
+      .on('end', async (rowCount: number) => {
+        try {
+          if (rowCount > MAX_ROW_COUNT) {
+            throw new BadRequestException('Row count cannot be greater than 500');
+          }
+
+          if (invalidRows.length) {
+            throw new BadRequestException(
+              `Please fix row number${invalidRows.length > 1 ? 's' : ''}: ${invalidRows.join(
+                ', '
+              )}. No users were uploaded`
+            );
+          }
+
+          if (invalidGroups.length) {
+            throw new BadRequestException(`Group ${invalidGroups.join(', ')} doesn't exist. No users were uploaded`);
+          }
+
+          if (existingUsers.length) {
+            throw new BadRequestException(
+              `User with email ${existingUsers.join(', ')} already exists. No users were uploaded`
+            );
+          }
+
+          for (let i = 0; i < users.length; i++) {
+            await this.inviteNewUser(currentUser, users[i]);
```

Comment:
This should be synchronous operation? and we should pass the entity manager.

Reply:
Since we decided to limit the number of rows to 500 we concluded that we can keep this synchronous. The problem with keeping this as an asynchronous insertion is that the pool size will be exhausted and it will fail. `await` makes sure that the previous insertion is done and sitting idle it will reuse that connection only. Let me know if any other options are suitable here. Thanks

Reply:
I will add the entitymanager here and check if its using the same connection from pool or not while keeping it asynchronous.